smithy-openapi/src/main/java/software/amazon/smithy/openapi/fromsmithy/protocols/AbstractRestProtocol.java

```diff
@@ -436,9 +437,97 @@ private void updateResponsesMapWithResponseStatusAndObject(
     ) {
         Shape operationOrError = shape.hasTrait(ErrorTrait.class) ? shape : operation;
         String statusCode = context.getOpenApiProtocol().getOperationResponseStatusCode(context, operationOrError);
-        ResponseObject response = createResponse(
+        ResponseObject newResponse = createResponse(
                 context, bindingIndex, eventStreamIndex, statusCode, operationOrError);
-        responses.put(statusCode, response);
+
+        // handle error code collision
+        if (responses.containsKey(statusCode)) {
+            newResponse =
+                    handleErrorCodeCollision(context.getConfig().isDisableOneOf(), statusCode, responses, newResponse);
+        }
+
+        responses.put(statusCode, newResponse);
+    }
+
+    private ResponseObject handleErrorCodeCollision(boolean disableOneOf,
```

Comment:
I think we can improve this if we handle this case in a different code path instead of patching up whatever we had before, that will make the code a lot easier to understand and maintain in the long run. In particular since we need to handle correctly the following cases:

1. Properties with `@HttpHeader` and `@required`: given that we will create a response having both responses those cannot be "required" anymore, thus we will need to drop it from the output (I tested this with your change and the required is emitted)
2. Responses with different Media Types, we cannot handle this case in particular and we need to error out instead of generating an incorrect response (I tested this with your change and the output is empty). I don't have a good story on how to do that, in particular an exception this deep in the code will lead to a bad UX, feel free to ask around about a good way to signal this issue to the client.

I think that we can simplify the code if we process the errors twice, one to aggregate them by status code and then we branch for the single response vs multi response cases. I think we can do that in the [`createResponses` method](https://github.com/awslabs/smithy/blob/main/smithy-openapi/src/main/java/software/amazon/smithy/openapi/fromsmithy/protocols/AbstractRestProtocol.java#L409), such that:

```
    private Map<String, ResponseObject> createResponses(
            Context<T> context,
            HttpBindingIndex bindingIndex,
            EventStreamIndex eventStreamIndex,
            OperationShape operation
    ) {
        Map<String, ResponseObject> result = new TreeMap<>();
        OperationIndex operationIndex = OperationIndex.of(context.getModel());

        StructureShape output = operationIndex.expectOutputShape(operation);
        updateResponsesMapWithResponseStatusAndObject(
                context, bindingIndex, eventStreamIndex, operation, output, result);

        Map<String, List<StructureShape>> errors = new LinkedHashMap<>();
        // Aggregate errors by status code to handle correctly those responses
        // that map to the same status
        for (StructureShape error : operationIndex.getErrors(operation)) {
            String statusCode = context.getOpenApiProtocol().getOperationResponseStatusCode(context, error);
            errors.computeIfAbsent(statusCode, (k) -> new ArrayList<>())
                    .add(error);
        }
        for (Map.Entry<String, List<StructureShape>> entry : errors.entrySet()) {
            List<StructureShape> shapes = entry.getValue();
            if (shapes.size() == 1) {
                updateResponsesMapWithResponseStatusAndObject(
                        context, bindingIndex, eventStreamIndex, operation, shapes.get(0), result);
            } else {
                updateResponsesMapWithResponseStatusAndMultipleObjects(
                        context, bindingIndex, eventStreamIndex, operation, shapes, result);
            }
        }
        return result;
    }

    // Handle the case with different responses for the same status code
    private void updateResponsesMapWithResponseStatusAndMultipleObjects(
            Context<T> context,
            HttpBindingIndex bindingIndex,
            EventStreamIndex eventStreamIndex,
            OperationShape operation,
            List<StructureShape> shapes,
            Map<String, ResponseObject> responses
    ) {
        // TODO
    }
```

Reply:
> 2\. Responses with different Media Types, we cannot handle this case in particular and we need to error out instead of generating an incorrect response (I tested this with your change and the output is empty). I don't have a good story on how to do that, in particular an exception this deep in the code will lead to a bad UX, feel free to ask around about a good way to signal this issue to the client.

I was confused, we can actually handle those using different media types keys, e.g.,

```
 "429": {
        ⋯
        "content": {
              "application/json": { ⋯ }
              "text/plain": { ⋯ }
```